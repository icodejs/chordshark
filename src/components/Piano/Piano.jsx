import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { toMidi } from '@tonaljs/midi';

import DeviceSelection from '../../components/DeviceSelection';

import { getDevice, setDeviceOctaves } from '../../client/localStorage';
import { flatNotes, flatToSharp, octavesOptions } from '../../config';
import { addNote, removeNote, containsNote } from '../../utils/notes';
import './Piano.scss';

const KEYS_PER_OCTAVE = 12;

const createPianoKeyEvent = (note, pianoOctave) => {
  const id = note + pianoOctave;

  return {
    id,
    midiNote: toMidi(id),
    timestamp: new Date().getTime(),
    target: 'Mouse click',
    rawVelocity: 127,
    channel: 1,
    octave: pianoOctave,
    number: toMidi(id),
    name: note,
  };
};

const parseNote = event => ({
  id: event.note.name + event.note.octave,
  midiNote: event.note.number,
  rawVelocity: event.rawVelocity,
  timestamp: event.timestamp,
  target: event.target,
  channel: event.channel,
  ...event.note,
});

class Piano extends Component {
  constructor(props) {
    super(props);
    this.state = {
      deviceSet: false,
      notesPressed: [],
    };
  }

  onNoteOn = e => {
    const note = parseNote(e);
    this.props.onNoteOn(note);
  };

  onNoteOff = e => {
    const note = parseNote(e);
    this.props.onNoteOff(note);
  };

  componentDidUpdate() {
    const { midiInputDevice } = this.props;
    const { deviceSet } = this.state;

    if (midiInputDevice && !deviceSet) {
      this.setState({ deviceSet: true }, () => {
        midiInputDevice.addListener('noteon', 'all', this.onNoteOn);
        midiInputDevice.addListener('noteoff', 'all', this.onNoteOff);
      });

      const { octaves } = getDevice();

      if (!octaves) {
        return;
      }

      this.props.selectNumberOfKeyboardOctaves({
        numberOfKeyboardOctaves: octaves,
      });
    }
  }

  componentWillUnmount() {
    const { midiInputDevice } = this.props;

    if (midiInputDevice) {
      midiInputDevice.removeListener('noteon', 'all', this.onNoteOn);
      midiInputDevice.removeListener('noteoff', 'all', this.onNoteOff);
    }
  }

  handleKeyboardKeysSelection = event => {
    const { value: numberOfKeyboardOctaves } = event.target;
    const octaves = parseInt(numberOfKeyboardOctaves, 10);

    setDeviceOctaves(octaves);

    this.props.selectNumberOfKeyboardOctaves({
      numberOfKeyboardOctaves: octaves,
    });
  };

  handleNoteClicked = (note, pianoOctave) => e => {
    e.preventDefault();
    const event = createPianoKeyEvent(note, pianoOctave);
    this.toggleNote(event);
    this.props.onNoteClick(event);
  };

  toggleNote(note) {
    if (containsNote(this.state.notesPressed, note.id)) {
      this.props.onNoteOff(note);

      this.setState({
        notesPressed: removeNote(this.state.notesPressed)(note.id),
      });
    } else {
      this.props.onNoteOn(note);

      this.setState({
        notesPressed: addNote(this.state.notesPressed)(note),
      });
    }
  }

  renderKeyboardKeysSelector = () => (
    <form className="octave-selector">
      <label>
        <select
          value={this.props.numberOfKeyboardOctaves}
          onChange={this.handleKeyboardKeysSelection}
        >
          {octavesOptions.map(({ keys, octaves }) => {
            return (
              <option key={octaves} value={octaves}>
                {`${keys}-key keyboard`}
              </option>
            );
          })}
        </select>
      </label>
    </form>
  );

  renderDeviceSelector = () => {
    const { midiInputs, handleDeviceSelection, selectedDevice } = this.props;
    return (
      <DeviceSelection
        {...{
          midiInputs,
          selectedDevice,
          onDeviceSelection: handleDeviceSelection,
        }}
      />
    );
  };

  renderNote = pianoOctave => (k, noteIndex) => {
    const { notesPressed, guideNotes } = this.props;
    const note = flatNotes[noteIndex];
    const id = `${note}_${pianoOctave}`;
    const selected = notesPressed.find(({ name, octave }) => {
      return flatToSharp(name) === flatToSharp(note) && octave === pianoOctave;
    });

    const scaleGuideNote = guideNotes.chord.find((n, i) => {
      const [noteName, noteOctave] = n.split(/(\d)/).filter(a => a);

      return (
        flatToSharp(noteName) === flatToSharp(note) &&
        parseInt(pianoOctave, 10) === parseInt(noteOctave, 10)
      );
    });

    return (
      <li
        key={id}
        onClick={this.handleNoteClicked(note, pianoOctave)}
        className={classNames(id, {
          selected,
          'scale-guide-note': scaleGuideNote,
        })}
      />
    );
  };

  renderPiano() {
    const { numberOfKeyboardOctaves } = this.props;
    const octavesOption = octavesOptions.find(
      ({ octaves }) => octaves === numberOfKeyboardOctaves,
    );

    return (
      <div className="piano">
        {[...Array(numberOfKeyboardOctaves)].map((o, index) => {
          const pianoOctave = index + octavesOption.startingOctave;

          return (
            <ul
              key={`octave-${pianoOctave}`}
              className={`octave-${pianoOctave}`}
            >
              {[...Array(KEYS_PER_OCTAVE)].map(this.renderNote(pianoOctave))}
            </ul>
          );
        })}
      </div>
    );
  }

  render() {
    return (
      <Fragment>
        {this.renderDeviceSelector()}
        {this.renderKeyboardKeysSelector()}
        {this.renderPiano()}
      </Fragment>
    );
  }
}

Piano.propTypes = {
  onNoteOn: PropTypes.func.isRequired,
  onNoteOff: PropTypes.func.isRequired,
  numberOfKeyboardOctaves: PropTypes.number.isRequired,
  selectNumberOfKeyboardOctaves: PropTypes.func.isRequired,
  onNoteClick: PropTypes.func,
  midiInputDevice: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  notesPressed: PropTypes.arrayOf(PropTypes.object),
  guideNotes: PropTypes.shape({
    scale: PropTypes.array,
    chord: PropTypes.array,
  }),
};

Piano.defaultProps = {
  notesPressed: [],
  guideNotes: { scale: [], chord: [] },
  numberOfKeyboardOctaves: 2,
  onNoteClick: () => {},
};

export default Piano;
