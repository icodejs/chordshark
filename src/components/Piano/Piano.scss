$br: 3px;
$grow-margin: -4.9%;
$bg: #282c34;

.black-key {
  background: #333;
  position: relative;
  height: 70%;
  flex-grow: 1;
  box-shadow: 0px 2px 1px 1px rgba(0, 0, 0, 0.2),
    0px -4px 1px 4px rgba(0, 0, 0, 0.5) inset;
  &:active {
    background: #282c34;
    box-shadow: 0px -4px 1px 4px rgba(0, 0, 0, 0.5) inset;
  }
  &.selected {
    background: #61dafb;
  }
}
.prev-is-black {
  margin-left: $grow-margin;
}
.next-is-black {
  margin-right: $grow-margin;
}

.notes-pressed {
  display: inline-block;
  padding-right: 10px;
}

.piano {
  background: $bg;
  height: 80vh;
  display: flex;
  justify-content: space-around;
  flex-direction: row;
  box-sizing: border-box;
  padding: 10px;

  ul {
    display: flex;
    flex-grow: 1;
    padding: 0;
    margin-left: -1px;
    &:first-child {
      margin-left: 0;
    }
  }
  li {
    flex-grow: 1.5;
    background: white;
    list-style: none;
    border-radius: 0 0 $br $br;
    border-right: 1px solid $bg;
    border-left: 1px solid $bg;
    margin-left: -1px;
    cursor: pointer;
    box-shadow: 0px -2px 1px 1px rgba(0, 0, 0, 0.2) inset;
    &:active {
      background: #eee;
    }
    &.scale-guide-note {
      background: yellow;
    }
    &.selected {
      background: #61dafb;
    }
    &:first-child {
      margin-left: 0 !important;
    }
    &:last-child {
      margin-right: 0 !important;
    }

    // Black Keys
    &:nth-child(2),
    &:nth-child(4),
    &:nth-child(7),
    &:nth-child(9),
    &:nth-child(11) {
      @extend .black-key;
    }

    // Prev key is black
    // 3, 5, 8, 10, 12
    &:nth-child(1),
    &:nth-child(3),
    &:nth-child(5),
    &:nth-child(8),
    &:nth-child(10),
    &:nth-child(12) {
      @extend .prev-is-black;
    }
    // Next key is black
    // 1, 3, 6, 8, 10
    &:nth-child(1),
    &:nth-child(3),
    &:nth-child(6),
    &:nth-child(8),
    &:nth-child(10) {
      @extend .next-is-black;
    }
  }
}
