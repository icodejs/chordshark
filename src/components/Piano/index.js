import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Piano from './Piano';
import * as actions from '../../common/actions';

const mapStateToProps = ({
  app: { numberOfKeyboardOctaves, selectedDevice, webMidiSupported },
}) => ({
  numberOfKeyboardOctaves,
  selectedDevice,
  webMidiSupported,
});

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Piano);
