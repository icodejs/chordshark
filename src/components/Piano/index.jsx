import React, { Component, Fragment } from 'react';
import classNames from 'classnames';
import { flatNotes, flatToSharp } from '../../config';
import './Piano.scss';

class Piano extends Component {
  constructor(props) {
    super(props);
    this.state = {
      deviceSet: false
    };
  }

  parseNote(event) {
    return {
      id: event.note.name + event.note.octave,
      midiNote: event.note.number,
      rawVelocity: event.rawVelocity,
      ...event.note
    };
  }

  onNoteOn = e => {
    const note = this.parseNote(e);
    this.props.onNoteOn(note);
  };

  onNoteOff = e => {
    const note = this.parseNote(e);
    this.props.onNoteOff(note);
  };

  componentDidUpdate() {
    const { midiInputDevice } = this.props;
    const { deviceSet } = this.state;

    if (midiInputDevice && !deviceSet) {
      this.setState({ deviceSet: true }, () => {
        midiInputDevice.addListener('noteon', 'all', this.onNoteOn);
        midiInputDevice.addListener('noteoff', 'all', this.onNoteOff);
      });
    }
  }

  componentWillUnmount() {
    const { midiInputDevice } = this.props;

    if (midiInputDevice) {
      midiInputDevice.removeListener('noteon', 'all', this.onNoteOn);
      midiInputDevice.removeListener('noteoff', 'all', this.onNoteOff);
    }
  }

  renderPiano() {
    const octaves = 2;
    return [...Array(octaves)].map((o, index) => {
      const pianoOctave = index + 3;

      return (
        <ul key={`octave-${pianoOctave}`} className={`octave-${pianoOctave}`}>
          {[...Array(12)].map((k, noteIndex) => {
            const note = flatNotes[noteIndex];

            const selected = this.props.notesPressed.find(({ name, octave }) => {
              return name === flatToSharp(note) && octave === pianoOctave;
            });

            return (
              <li
                key={`${note}_${pianoOctave}`}
                className={classNames(`${note}_${pianoOctave}`, { selected })}
              />
            );
          })}
        </ul>
      );
    });
  }

  render() {
    return (
      <Fragment>
        <div className="piano">{this.renderPiano()}</div>
      </Fragment>
    );
  }
}

export default Piano;
