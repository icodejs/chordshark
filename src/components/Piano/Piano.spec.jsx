import React from 'react';
import renderer from 'react-test-renderer';

import Piano from './Piano';

describe('Piano', () => {
  describe('render', () => {
    it('should render component', () => {
      const props = {
        onNoteOn: () => {},
        onNoteOff: () => {},
        selectNumberOfKeyboardOctaves: () => {},
        handleDeviceSelection: () => {},
        midiInputs: [
          {
            id: 'fake-id1',
            manufacturer: 'fake-manufacturer',
            name: 'fake-name',
            type: 'fake-type',
          },
          {
            id: 'fake-id2',
            manufacturer: 'fake-manufacturer',
            name: 'fake-name',
            type: 'fake-type',
          },
        ],
      };
      const json = renderer.create(<Piano {...props} />).toJSON();
      expect(json).toMatchSnapshot();
    });
  });
});
