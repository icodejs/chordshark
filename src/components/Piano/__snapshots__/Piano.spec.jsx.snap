// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piano render should render component 1`] = `
Array [
  <form
    className="device-selection"
  >
    <label>
      <select
        onChange={[Function]}
      >
        <option
          value="-1"
        >
          Choose device
        </option>
        <option
          value="fake-name"
        >
          fake-manufacturer fake-name
        </option>
        <option
          value="fake-name"
        >
          fake-manufacturer fake-name
        </option>
      </select>
    </label>
  </form>,
  <form
    className="octave-selector"
  >
    <label>
      <select
        onChange={[Function]}
        value={2}
      >
        <option
          value={2}
        >
          25-key keyboard
        </option>
        <option
          value={4}
        >
          49-key keyboard
        </option>
        <option
          value={5}
        >
          61-key keyboard
        </option>
        <option
          value={6}
        >
          76-key keyboard
        </option>
        <option
          value={7}
        >
          88-key keyboard
        </option>
      </select>
    </label>
  </form>,
  <div
    className="piano"
  >
    <ul
      className="octave-3"
    >
      <li
        className="C_3"
        onClick={[Function]}
      />
      <li
        className="Db_3"
        onClick={[Function]}
      />
      <li
        className="D_3"
        onClick={[Function]}
      />
      <li
        className="Eb_3"
        onClick={[Function]}
      />
      <li
        className="E_3"
        onClick={[Function]}
      />
      <li
        className="F_3"
        onClick={[Function]}
      />
      <li
        className="Gb_3"
        onClick={[Function]}
      />
      <li
        className="G_3"
        onClick={[Function]}
      />
      <li
        className="Ab_3"
        onClick={[Function]}
      />
      <li
        className="A_3"
        onClick={[Function]}
      />
      <li
        className="Bb_3"
        onClick={[Function]}
      />
      <li
        className="B_3"
        onClick={[Function]}
      />
    </ul>
    <ul
      className="octave-4"
    >
      <li
        className="C_4"
        onClick={[Function]}
      />
      <li
        className="Db_4"
        onClick={[Function]}
      />
      <li
        className="D_4"
        onClick={[Function]}
      />
      <li
        className="Eb_4"
        onClick={[Function]}
      />
      <li
        className="E_4"
        onClick={[Function]}
      />
      <li
        className="F_4"
        onClick={[Function]}
      />
      <li
        className="Gb_4"
        onClick={[Function]}
      />
      <li
        className="G_4"
        onClick={[Function]}
      />
      <li
        className="Ab_4"
        onClick={[Function]}
      />
      <li
        className="A_4"
        onClick={[Function]}
      />
      <li
        className="Bb_4"
        onClick={[Function]}
      />
      <li
        className="B_4"
        onClick={[Function]}
      />
    </ul>
  </div>,
]
`;
